File: src/test/java/me/nasukhov/bot/command/LearnWordTest.java
package me.nasukhov.bot.command;

import me.nasukhov.bot.io.Channel;
import me.nasukhov.bot.io.Input;
import me.nasukhov.bot.io.Output;
import me.nasukhov.bot.io.User;
import me.nasukhov.dictionary.DictionaryRepository;
import me.nasukhov.dictionary.PartOfSpeech;
import me.nasukhov.dictionary.Word;
import me.nasukhov.study.Group;
import me.nasukhov.study.ProgressRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.util.ArrayList;
import java.util.List;

import static org.mockito.Mockito.*;

public class LearnWordTest {
    private static final int MAX_AMOUNT_OF_WORDS_PER_REQUEST = 3;

    private LearnWord handler;
    private DictionaryRepository dictionaryRepository;
    private ProgressRepository progressRepository;
    private Output output;

    @BeforeEach
    void setup() {
        output = mock(Output.class);
        dictionaryRepository = mock(DictionaryRepository.class);
        progressRepository = mock(ProgressRepository.class);
        handler = new LearnWord(dictionaryRepository, progressRepository);
    }

    @Test
    void testHandleUnsupportedCommand() {
        Input input = new Input(
                "Some input",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        handler.handle(input, output);

        verifyNoInteractions(output, dictionaryRepository, progressRepository);
    }

    @ParameterizedTest
    @ValueSource(strings = {"/learn", "/learn@botName"})
    public void testLearnWordsWhenThereAreNoneLeft(String rawInput) {
        Input input = new Input(
                rawInput,
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        Group group = new Group(input.channel().id);

        int lastLearnedWordId = 123;
        when(progressRepository.getLastLearnedWordId(group)).thenReturn(lastLearnedWordId);
        when(dictionaryRepository.getChunk(MAX_AMOUNT_OF_WORDS_PER_REQUEST, lastLearnedWordId)).thenReturn(new ArrayList<>());

        handler.handle(input, output);

        verify(output).write("Вы изучили все слова из нашего словаря - больше новых слов нет.");
        verify(progressRepository).getLastLearnedWordId(group);
        verifyNoMoreInteractions(progressRepository);
    }

    @ParameterizedTest
    @ValueSource(strings = {"/learn", "/learn@botName"})
    public void testLearnWords(String rawInput) {
        Input input = new Input(
                rawInput,
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        Group group = new Group(input.channel().id);
        int lastLearnedWordId = 0;
        when(progressRepository.getLastLearnedWordId(group)).thenReturn(lastLearnedWordId);
        when(dictionaryRepository.getChunk(MAX_AMOUNT_OF_WORDS_PER_REQUEST, lastLearnedWordId)).thenReturn(new ArrayList<>(){{
            add(new Word(124, "Word1", "Translation1", "Description1", PartOfSpeech.NOUN));
            add(new Word(125, "Word2", "Translation2", "Description2", PartOfSpeech.NOUN));
        }});

        handler.handle(input, output);

        verify(output).write("Word1 - Translation1\n\nWord2 - Translation2");
        verify(progressRepository).getLastLearnedWordId(group);
        verify(progressRepository).setLastLearnedWords(group, List.of(124, 125));
        verifyNoMoreInteractions(progressRepository);
    }
}

File: src/test/kotlin/me/nasukhov/tukitalearner/bot/LearnWordTest.kt
package me.nasukhov.tukitalearner.bot

import me.nasukhov.bot.io.Channel
import me.nasukhov.bot.io.Input
import me.nasukhov.bot.io.NoOutput
import me.nasukhov.bot.io.TextOutput
import me.nasukhov.bot.io.User
import me.nasukhov.tukitalearner.dictionary.DictionaryEntry
import me.nasukhov.tukitalearner.dictionary.DictionaryRepository
import me.nasukhov.tukitalearner.study.Group
import me.nasukhov.tukitalearner.study.LearnedResourceRepository
import me.nasukhov.tukitalearner.study.ProgressTracker
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertInstanceOf
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional
import kotlin.test.assertEquals

@SpringBootTest
@Transactional
class LearnWordTest {
    @Autowired
    private lateinit var handler: LearnWord

    @Autowired
    private lateinit var dictionary: DictionaryRepository

    @Autowired
    private lateinit var progressTracker: ProgressTracker

    @Autowired
    private lateinit var learnedResourceRepository: LearnedResourceRepository

    @BeforeEach
    fun setup() {
        dictionary.saveAll(
            listOf(
                DictionaryEntry("гьой", "собака", "Делает кусь"),
                DictionaryEntry("кету", "кошка", "Для глажки"),
                DictionaryEntry("рухья", "дерево", "Вырабатывает кислород ми спасает от жары"),
            ),
        )
    }

    @AfterEach
    fun reset() {
        learnedResourceRepository.deleteAll()
        dictionary.deleteAll()
    }

    @Test
    fun handleUnsupportedInput() {
        val channel = Channel("SomeChannelId")
        val user = User("SomeId", "SomeName")
        val input = Input("learn", channel, user)

        val output = handler.handle(input)
        assertInstanceOf<NoOutput>(output)
    }

    @Test
    fun handleWhenAllWordsAreLearned() {
        val channel = Channel("SomeChannelId")
        val user = User("SomeId", "SomeName")
        val input = Input("/learn", channel, user)

        progressTracker.setLastLearnedWords(Group(channel.id), dictionary.findAll())

        val output = handler.handle(input)
        assertInstanceOf<TextOutput>(output)
        assertEquals("Вы изучили все слова из нашего словаря - больше новых слов нет.", output.value)
    }

    @Test
    fun handleNormally() {
        val channel = Channel("SomeChannelId")
        val user = User("SomeId", "SomeName")
        val input = Input("/learn", channel, user)

        val output = handler.handle(input)
        assertInstanceOf<TextOutput>(output)

        assertEquals(
            """
            гьой - собака

            кету - кошка

            рухья - дерево
            """.trimIndent(),
            output.value,
        )
    }
}
